pkg/log/klog/klog.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package klog contains logging related functionality
+package klog
+
+import (
+	"k8s.io/klog"
```

Comment:
the old PR that i did made /pkg logger agnostic, leaving it to the front end to choose its logger.
are there plans to make /pkg to not import `"k8s.io/klog"`?

Reply:
this package should only be imported by `cmd/`, (that's buried in the wall of text in the OP)

Reply:
`pkg/` should use `pkg/log` functions against the private `pkg/log.std` which can be set with `pkg/log.SetDefault`

I think we will need to introduce a no-op default logger that is the actual default until `SetDefault` is called

Reply:
> Having a shim means users do not have to use klog, if they're happy with klog they just need to import and call sigs.k8s.io/kind/pkg/log/klog.Use() somewhere, otherwise they can implement the little log interface and pass it to sigs.k8s.io/kind/pkg/log.SetDefault(). klog is standard in the kubernetes ecosystem and V() logging is nice, but we can be flexible.

if kind 's `pkg` imports `klog`, and in a case where a third-party front end uses kind as a library, this:
- could create a conflict if the third-party logger already depends on a different klog version for one reason or another. i think this will force the front-end to shift to the particular klog version that kind uses.
- will result in a couple of loggers to be imported if the third-party front-end already uses a separate logger. the klog import will not be needed in this case.

i would have moved the klog import to `cmd` and document (e.g. in a document called "kind as a library"), how to setup the logger if one wishes to import `kind/pkg`.

Reply:
> if kind 's pkg imports klog, and in a case where a third-party front end uses kind as a library, this:
> - could create a conflict if the third-party front end already depends on a different klog version for one reason or another. i think this will force the front-end to shift to the particular klog version that kind uses.

How will this be any different, this is the same go module?

> - will result in a couple of loggers to be imported if the third-party front-end already uses a separate logger. the klog import will not be needed in this case.

No? None of the code they link against will import `pkg/log/klog` unless they do themselves or they import `cmd/`.

> i would have moved the klog import to cmd and document (e.g. in a document called "kind as a library"), how to setup the logger if one wishes to import kind/pkg.

 we need an implementation of the shim against klog somwhere, that's all this is. cmd will be the only one importing `pkg/log/klog` and calling `Use()`

Reply:
> How will this be any different, this is the same go module?

as an example, the cluster-api folk are currently trying to create a kind based deployer where the apimachinery kind uses is not suitable for their needs. they have another version in the go.mod.
and technically the versions are conflicting.

@chuckha has more context.

> No? None of the code they link against will import pkg/log/klog unless they do themselves or they import cmd/.

but klog would still be present in the go.mod, unless i'm mistaken here?

Reply:
The reason the klog shim implementation is not in `cmd/` is I actually expect most library users will be using klog/glog because of the Kubernetes ecosystem. They're meant to be able to use this implementation, but they probably should not import cmd/ vs just exec-ing to a kind binary.

Since `pkg/*` doesn't import klog* (ONLY `pkg/log/klog`) you only transitively bring klog into your binary if you import `pkg/log/klog` specifically (or `cmd/`). Otherwise you just get the very-klog-like shim and you'll need to figure out implementing it against some other logger.

Because of our existing dependencies, `klog` is already in `go.mod` and has been. (I think this is due to kustomize, and I have seperate thoughts on fixing that, unrelated to this PR :sweat_smile:)

Reply:
To summarize:
- klog will be in your dependencies if you vendor / depend on kind, and _already is prior to this PR_.
- klog will not actually be linked into your binary unless you import the shim package*
- cmd/ will import and link against klog
- logging will look pretty similar to kubeadm

\* except that it is because one of our deps, will fix that later

Reply:
> as an example, the cluster-api folk are currently trying to create a kind based deployer where the apimachinery kind uses is not suitable for their needs. they have another version in the go.mod.
> and technically the versions are conflicting.

This is already an issue and already solved by:
https://github.com/golang/go/wiki/Modules#when-should-i-use-the-replace-directive

Reply:
in the cluster-api `go.mod`:
`replace k8s.io/klog => k8s.io/klog $VERSION_CLUSTER_API_WANTS`

Reply:
The shim will make this easier to do safely because it makes `pkg/` (again, not the shim package, but no other pkg/ will import that) depend on a tiny subset of klog's functionality that exists in all known versions. Seperately we need to root out the dependency caused by the patch handling code, which will happen but is a little more involved :sweat_smile:

Reply:
wanted @chuckha to leave a comment about the kind provider use case, but feel free to `hold cancel` to unblock the follow ups.